### DISCUSSION TITLE
What's the difference between a testResultsProcessor and a reporter?

### DISCUSSION THREAD

**Discussion by SevenOutman** [jest-community/jest-junit#158] [Q&A] [answered]
What's different when I use jest-junit as a reporter v.s. a testResultsProcessor?

**Comment by palmerj3** [comment, marked as answer]
Very little. But the subtle difference is that a testResultsProcessor only receives details about test runs after all tests are finished.

Whereas a reporter has the ability to receive test results after individual tests and/or test suites are finished.

testResultsProcessor remains in jest because Facebook needs it internally (or so I've been told).

**Comment by SevenOutman** [comment]
So if I just want to generate a JUnit style report after all tests are finished (in my case for GitLab CI integration), it's merely no difference to use jest-junit as a testResultsProcessor or as a reporter.


And when I need it to report dynamically after each test case (e.g. IDE integrations), I should only use it as a reporter.


Am I getting it correct?













At 2021-02-22 17:55:30, "Jason Palmer" <notifications@github.com> wrote:

Very little. But the subtle difference is that a testResultsProcessor only receives details about test runs after all tests are finished.

Whereas a reporter has the ability to receive test results after individual tests and/or test suites are finished.

testResultsProcessor remains in jest because Facebook needs it internally (or so I've been told).

**Comment by palmerj3** [comment]
You should use it as a reporter. testResultsProcessor support is only there for legacy reasons but it could be removed in the future.

jest-junit only writes junit files at the end of a test run anyway.

**Reply by fatso83** [reply]
I think this could be better communicated in the docs, maybe even linking to this discussion 😄 I was wondering why the config only took effect when using it as a reporter, as opposed to configuring it as a testResultsProcessor.

**Reply by SimenB** [reply]
`reporter` is mentioned first, but I agree a sentence saying that using it as `testResultsProcessor` is discouraged is probably a good idea. PR? 😀

**Comment by palmerj3** [comment]
Using it as a reporter allows you a lot more flexibility in the future should you decide to use some of jest-junit's configuration options

**Reply by SevenOutman** [reply]
Thank you for the clear answers!

**Comment by zoltan-boros** [comment]
Back to the original question:

> _What's the difference between a testResultsProcessor and a reporter?_

As per [Jest API](https://jestjs.io/docs/configuration) `testResultsProcessor` is string, whereas `reporters` is array. This implies that only one test result processor can be used at a time, whereas multiple reporters simultaneously. At least this is how I interpret the API.
